tests/test_record_cylc_install_options.py

```diff
@@ -0,0 +1,186 @@
+    install_filepath = tmp_path / 'rose-suite-cylc-install.conf'
+    rose_conf_filepath = tmp_path / 'rose-suite.conf'
+    install_filepath.write_text(
+        "!opts=A B"
+    )
+    rose_conf_filepath.write_text(
+        "opts=C D"
+    )
+    opts = SimpleNamespace(
+        opt_confs='E F', defines=[], define_suites=[]
+    )
+    record_cylc_install_options(opts=opts, dest_root=tmp_path)
+    assert install_filepath.read_text() == \
+        (
+            "# This file records CLI Options.\n"
+            "\n!opts=A B E F\n"
+        )
+    assert rose_conf_filepath.read_text() == \
+        (
+            '# Config Options \'C D A B E F\' '
+            'from CLI appended to options already in '
+            '`rose-suite.conf`.\n'
+            '!opts=C D A B E F\n'
+        )
+
+
+@pytest.mark.parametrize(
+    (
+        'opts, rose_suite_conf, install_file,env_inserts,'
+        'kgo_install_file, kgo_rose_file'
+    ),
+    [
+        # ``cylc install -D '[env]FOO=1' -S 'X=Y'``
+        (
+            # opts
+            SimpleNamespace(
+                opt_confs='', defines=['[env]FOO=1'], define_suites=['X=Y']
+            ),
+            # rose-suite.conf
+            'opts=foo',
+            # Existing rose-suite-cylc-install.conf
+            '',
+            # No environment vars set
+            {},
+            # Expected rose-suite-cylc-install.conf
+            '[env]\nFOO=1\n[jinja2:suite.rc]\nX=Y\n',
+            # Expected rose-suite.conf (Only need opts section).
+            'opts=foo'
+        ),
+        # Clean install with ROSE_SUITE_OPT_CONF_KEYS set
+        (
+            # opts
+            SimpleNamespace(
+                opt_confs='c', defines=[], define_suites=[]
+            ),
+            # rose-suite.conf
+            'opts=a',
+            # Existing rose-suite-cylc-install.conf
+            '',
+            # No environment vars set
+            {'ROSE_SUITE_OPT_CONF_KEYS': 'b'},
+            # Expected rose-suite-cylc-install.conf
+            '!opts=b c\n',
+            # Expected rose-suite.conf (Only need opts section).
+            'opts=a b c'
+        ),
+        # re-install with ROSE_SUITE_OPT_CONF_KEYS set
+        (
+            # opts
+            SimpleNamespace(
+                opt_confs='d', defines=[], define_suites=[]
+            ),
+            # rose-suite.conf
+            'opts=a',
+            # Existing rose-suite-cylc-install.conf
+            '!opts=b c\n',
+            # No environment vars set
+            {'ROSE_SUITE_OPT_CONF_KEYS': 'b'},
+            # Expected rose-suite-cylc-install.conf
+            '!opts=b c d\n',
+            # Expected rose-suite.conf (Only need opts section).
+            'opts=a b c d'
+        )
+    ]
+)
+def test_functional_record_cylc_install_options(
+    monkeypatch, tmp_path, opts, rose_suite_conf, install_file, env_inserts,
+    kgo_install_file, kgo_rose_file
+):
+    class MockConfigLoader(ConfigLoader):
```

Comment:
@oliver-sanders  I have become concerned that given that this test is basically a functional test of one of the outward facing functions that this mocking approach will become very messy once we start adding (for example) `log` files.
Is it OK if I roll back on it and just touch the file system.

Reply:
In order to use the rose Config API would need to pull the CLI and API parts into separate functions.

Here's a quick attempt, I took the `record_cylc_install_options` function and re-ordered the different bits like so:

* load from FS
* process
* dump to FS

The process bit is the bit you want to test.

```python
def record_cylc_install_options(
    dest_root=None,
    opts=None,
    dir_=None,
):
    # LOAD FILES
    opt_conf_filepath = Path(dest_root) / 'rose-suite-cylc-install.conf'
    rose_conf_filepath = Path(dest_root) / 'rose-suite.conf'
    dumper = ConfigDumper()
    loader = ConfigLoader()

    new_opt_config = get_cli_opts_node(opts)

    if opt_conf_filepath.is_file():
        old_opt_config = loader.load(str(opt_conf_filepath))
    else:
        old_opt_config = ConfigNode()

    if not rose_conf_filepath.is_file():
        rose_conf_filepath.touch()
    rose_suite_conf = loader.load(str(rose_conf_filepath))

    # PROCESS FILES
    new_opt_config, rose_suite_conf = _record_cylc_install_options(
        old_opt_config,
        new_opt_config,
        rose_suite_conf
    )

    # DUMP FILES
    dumper.dump(new_opt_config, str(opt_conf_filepath))
    dumper(rose_suite_conf, rose_conf_filepath)

    return new_opt_config, rose_suite_conf['opts']


def _record_cylc_install_options(
    old_opt_config,
    new_opt_config,
    rose_suite_conf
):
    # THIS HAS NO FILESYSTEM LOGIC SO IS GOOD FOR TESTING
    if old_opt_config:  # TODO check this does what you want
        new_opt_config = merge_rose_cylc_suite_install_conf(
            old_opt_config,
            new_opt_config
        )
        new_opt_config.comments = [
            ' This file records CLI Options.'
        ]
    rose_suite_conf = get_installed_rose_suite_conf_node(
        rose_suite_conf,
        new_opt_config
    )
    return new_opt_config, rose_suite_conf
```